### DISCUSSION TITLE
Flutter_sound Recorder

### DISCUSSION THREAD

**Discussion by Stimm4** [Canardoux/flutter_sound#1081] [Q&A]
Hi,

I start to Flutter, but I need to record a sound with a specific bufferSize.

Using the example, this doesn't work for me.

Is it possibl to specify a bufferSize ?

Thanks

**Comment by Larpoux** [comment]
Which platform? Web, iOS or Android?
Which tool ? Recorder or player ?
Files, buffers, uri or streams?

**Reply by Stimm4** [reply]
I use it in Android for now, and I tried to set the BufferSize like this :


  `Future<void> _startRecording() async {
        if (await Permission.microphone.request().isGranted) {
          setState(() {
            _isRecording = true;
          });
     
        _recorder!.startRecorder(
          codec: Codec.pcm16,
          toStream: _audioStreamController.sink,
          bufferSize : 100, // This doesn't work
            bitRate : 8000
  
        );
  
        _audioStreamController.stream.listen((buffer) {
          _processBuffer(buffer);
        });
      } else {
        print('Microphone permission not granted');
      }
    }
  
    void _processBuffer(Uint8List buffer) {
      // Convert the Uint8List to Int16List if necessary and pass it to native code
      final Int16List int16Buffer = Int16List.view(buffer.buffer);
      _audioProcessor.processAudio(int16Buffer);
    }`

**Comment by Larpoux** [comment]
Oh, yes. Recorder

**Comment by Larpoux** [comment]
I am going to look to Flutter Sound code.
Btw: `bitRate`is not used in case of PCM. The correct parameter is `sampleRate`

How do you know that your parameter `bufferSize` does not work?

**Reply by Stimm4** [reply]
Yes, that was a test, I usually use sampleRate.

I used the function below, that print at each frames the size of the buffer.
And this size is equals to 355 or less sometimes. (Different to the desired value)

`
  void processAudio(Int16List buffer) {
      final Pointer<Int16> ptr = calloc(buffer.length);
      ptr.asTypedList(buffer.length).setAll(0, buffer);
      print('Process audio data! + buffer.length = '+(buffer.length).toString());
      // _processAudio(ptr, buffer.length);
      calloc.free(ptr);
  }
`

**Comment by Larpoux** [comment]
Here is the Flutter Sound code : 
```java
		int minBufferSize = AudioRecord.getMinBufferSize
			(
				sampleRate,
			      	channelConfig,
				tabCodec[codec.ordinal()]
			) * 2;

			int bufLn = Math.max(minBufferSize, bufferSize);
```

**Comment by Larpoux** [comment]
`bufferSize` variable is supposed to be your parameter.

**Reply by Stimm4** [reply]
I tried to set the bufferSize to 1200, but I got the same size (355).

If the code below :
int minBufferSize = AudioRecord.getMinBufferSize
  (
  sampleRate,
			     channelConfig,
  tabCodec[codec.ordinal()]
  ) * 2;

int bufLn = Math.max(minBufferSize, bufferSize);

If he size of the buffer returned is 355, that means, that I'm able to set it to a value higher than 355 ?

**Comment by Larpoux** [comment]
Things are not so simple 👀 
The variable `bufLn` that I computed is used : 
```java
				// gets the voice output from microphone to byte format
				if ( Build.VERSION.SDK_INT >= 23 )
				{
						n = recorder.read(byteBuffer.array(), 0, bufLn, AudioRecord.READ_NON_BLOCKING);

				}
				else
				{
						n = recorder.read(byteBuffer.array(), 0, bufLn);
				}
```
`bufLn` is the size of my buffer. But Android can give me a packet smaller.

**Comment by Larpoux** [comment]
because we don't want to BLOCK when Android >= 23

**Reply by Stimm4** [reply]
Ho Drat!
So Impossible to be sure to have the exact bufferSize set with READ_NON_BLOCKING.
Thank you for your response !
